src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -1454,6 +1454,102 @@ void MacroAssembler::update_word_crc32(Register crc, Register v, Register tmp1,
   xorr(crc, crc, tmp2);
 }
 
+
+// This improvement (vectorization) is based on java.base/share/native/libzip/zlib/zcrc32.c.
+// To make it, following steps are taken:
+//  1. in zcrc32.c, modify N to 16 and related code,
+//  2. re-generate the tables needed, we use tables of (N == 16, W == 4)
+//  3. finally vectorize the code (original implementation in zcrc32.c is just scalar code).
+// New tables for vector version is after table3.
+void MacroAssembler::vector_update_crc32(Register crc, Register buf, Register len, const int64_t unroll_words,
+                                         Register tmp1, Register tmp2, Register tmp3, Register tmp4,
+                                         Register table0, Register table3, const int64_t single_table_size) {
+    const int N = 16, W = 4;
+    const Register blks = tmp2;
+    const Register tmpTable = tmp3, tableN16 = tmp4;
+    const VectorRegister vcrc = v4, vword = v8, vtmp = v12;
+    Label VectorLoop;
+    Label LastBlock;
+
+    add(tableN16, table3, 1*single_table_size*sizeof(juint), tmp1);
+
+    if (MaxVectorSize == 16) {
+      vsetivli(zr, N, Assembler::e32, Assembler::m4, Assembler::ma, Assembler::ta);
+    } else if (MaxVectorSize == 32) {
+      vsetivli(zr, N, Assembler::e32, Assembler::m2, Assembler::ma, Assembler::ta);
+    } else {
+      assert(MaxVectorSize > 32, "sanity");
+      vsetivli(zr, N, Assembler::e32, Assembler::m1, Assembler::ma, Assembler::ta);
+    }
+    vmv_v_x(vcrc, zr);
+    zext_w(crc, crc);
```

Comment:
What is this 'zext_w' for? And I see following two instructions on `kernel_crc32` entry which I think already zero-extended 32-bit `crc`?

```
  mv(tmp5, right_32_bits);
  andn(crc, tmp5, crc);
```